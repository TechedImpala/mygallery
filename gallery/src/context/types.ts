import { ApplicationStateEnum } from "../types";

export interface IAppProviderState {
  activeState: ApplicationStateEnum;
  setActiveState: React.Dispatch<
    React.SetStateAction<ApplicationStateEnum>
  > | null;
  activeTopic: string;
  setActiveTopic: React.Dispatch<React.SetStateAction<string>> | null;
  menuVisible: boolean;
  setMenuVisibility: React.Dispatch<React.SetStateAction<boolean>> | null;
}
