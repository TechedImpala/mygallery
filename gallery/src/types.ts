export enum ApplicationStateEnum {
  GRID = "GRID",
  MENU = "MENU",
}

export interface IUrls {
  raw: string;
  full: string;
  regular: string;
  small: string;
  thumb: string;
  small_s3: string;
}
