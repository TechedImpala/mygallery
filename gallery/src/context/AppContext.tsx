import { createContext, useContext, useState } from "react";
import { ApplicationStateEnum } from "../types";
import { IAppProviderState } from "./types";

const AppContext = createContext<IAppProviderState>({
  activeState: ApplicationStateEnum.MENU,
  setActiveState: null,
  activeTopic: "",
  setActiveTopic: null,
  menuVisible: true,
  setMenuVisibility: null,
});

interface IAppProviderProps {
  children: React.ReactNode | React.ReactNode[];
}

const AppProvider = ({ children }: IAppProviderProps) => {
  const [menuVisible, setMenuVisibility] = useState(true);
  const [activeState, setActiveState] = useState<ApplicationStateEnum>(
    ApplicationStateEnum.MENU
  );

  const [activeTopic, setActiveTopic] = useState<string>("");

  const value = {
    activeState,
    setActiveState,
    activeTopic,
    setActiveTopic,
    menuVisible,
    setMenuVisibility,
  };

  return <AppContext.Provider value={value}>{children}</AppContext.Provider>;
};

export const useAppContext = () => useContext(AppContext);

export default AppProvider;
