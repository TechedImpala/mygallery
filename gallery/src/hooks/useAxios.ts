import { useState, useEffect, useCallback } from "react";
import axios, { AxiosRequestConfig, AxiosResponse } from "axios";

const baseUrl = process.env.API_BASE_URL || "https://api.unsplash.com";

export const axiosInstance = axios.create({
  baseURL: baseUrl,
  responseType: "json",
});

interface IUseAxiosProps extends AxiosRequestConfig {
  manual?: boolean;
}

const useAxios = <T>(requestConfig: IUseAxiosProps) => {
  const [data, setData] = useState<T | null>(null);
  const [loading, setLoading] = useState<boolean>(false);
  const [error, setError] = useState<unknown>(null);

  const { url } = requestConfig;

  const handleRequest = useCallback(() => {
    setLoading(true);
    axiosInstance
      .request(requestConfig)
      .then((response: AxiosResponse) => {
        setData(response.data);
        setLoading(false);
      })
      .catch((error: unknown) => {
        setError(error);
        setLoading(false);
      });
  }, [url]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    if (!requestConfig.manual) handleRequest();
  }, [requestConfig.manual, handleRequest]);

  return { data: data as T, loading, error, handleRequest };
};

export default useAxios;
