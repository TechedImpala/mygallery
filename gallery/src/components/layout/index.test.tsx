import React from "react";
import { render, screen } from "@testing-library/react";
import Layout from "./";
import { useAppContext } from "../../context/AppContext";
import { ApplicationStateEnum } from "../../types";
import useAxios from "../../hooks/useAxios";
import { topicResponse } from "./__fixtures___/index.fixtures";

jest.mock("../../context/AppContext");
jest.mock("../../hooks/useAxios");

describe("Layout:", () => {
  test("renders the menu and grid with no content", () => {
    (useAxios as jest.Mock)
      .mockReturnValueOnce({
        data: topicResponse,
        loading: false,
      })
      .mockReturnValue({
        data: null,
        loading: false,
        handleRequest: jest.fn(),
      });
    (useAppContext as jest.Mock).mockReturnValue({
      activeState: ApplicationStateEnum.MENU,
      activeTopic: "",
    });
    const { container, getByText } = render(<Layout />);

    expect(getByText("Please select a topic")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
