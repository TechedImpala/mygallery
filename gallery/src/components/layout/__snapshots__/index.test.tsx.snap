// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout: renders the menu and grid with no content 1`] = `
<div>
  <div
    class="sc-jqUVSM hBbAhQ"
  >
    <header
      class="sc-ftvSup icNAsx"
    >
      <div
        class="sc-papXJ iiKwxX"
        data-testid="menu-item-color-of-water"
      >
        Color of Water
      </div>
      <div
        class="sc-papXJ iiKwxX"
        data-testid="menu-item-people"
      >
        Person
      </div>
    </header>
    <div
      class="sc-gsnTZi kXspxg"
    >
      Please select a topic
    </div>
  </div>
</div>
`;
